### PR TITLE
#486-part-2-fix-review-restart-for-multiple-sections

### PR DESCRIPTION
**Pull request by andreievg**
Fixes: #486 

The solution didn't end up being a simple one, but the logic seems sound. 

The idea:

getFullReviewStructure has another parameter now, awaitMode, when set to true, query will be skipped. It will also return `getFullReviewStructureAsync`, when awaitMode is set to true, this method can be used like this `const structure = await getFullReviewStructureAsync()`

What happens ?

getFullReviewStructureAsync creates a promise, and sets the resolver to a state variable awaitModeResolver. It also sets awaitModeTrigger which initiates a useGetReviewResponsesQuery. 

When there is a error or when full structure is generated, awaitModeResolver is called, which in turn resets `awaitModeTrigger` (to stop any queries) and resolves new structure (or rejects). Btw `await getFullReviewStructureAsync` will wait for either the `resolve` or `reject` to be called in the promise that was created

### Testing

using master and back end branch `FE-helpers-for-bug-fixes-from-demo`, do ABC123 review, with LOQ and all other approved. Do applicant changes, and notice what happens when re-review is pressed (check other sections, you will see 'review' button instead of previous decisions)

do the same on this branch, you will note previous decisions for other sections still have the decision that was made

### CODE REVIEW

**Comment by andreievg** [comment]
Please hold of a review, I am not happy about one aspect of this.

**Comment by andreievg** [comment]
I had to restructure it a bit (I didn't like different behaviours within the same hook), even though restructure is a bit less dry, I think it's way more readable; 783c3a6d2bb36d62fcfeadc431bf331e56cb56a8

Basically separate into two hooks `useGetFullReviewStructure` and `useGetFullReviewStructureAsync`, with common helpers. And `skip` is now controlled by `awaitModeResolver` being null.

**Comment by andreievg** [comment]
even though I did `git mv` it seems git didn't see that useGetFullReviewStructure was moved ; /

**Comment by andreievg** [comment]
This kind of 'async' type option would be very useful for checking things before submitting (i.e. 'was the review locked while user was working on it ?` `was review set to pending while consolidator was working on it ?`

**Review by nmadruga — COMMENTED** [review]
Just made comments to remove prints so far. I think this one would be better to review after the merge with master is done, but after trying to do it it doesn't seem very easy without having to understand exactly the logic used for this. So I'll leave that until Andrei returns.

Pending: resolve conflicts with `master` and do more tests (I've done tests on it before the merge and seems good)

File: src/utils/hooks/useGetFullReviewStructure/useGetFullReviewStructureAsync.tsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      userId: currentUser?.userId as number,
+    },
+    fetchPolicy: 'network-only',
+    skip: !awaitModeResolver,
+  })
+
+  console.log(awaitModeResolver, !!awaitModeResolver)
+
+  useEffect(() => {
+    console.log('here')
```
[the remark this answers is not in the document]
Comment (andreievg) [reply src/utils/hooks/useGetFullReviewStructure/useGetFullReviewStructureAsync.tsx]: 1d66d8211e2de3fc110a0c0cf1c2715bb48ac649

File: src/utils/hooks/useGetFullReviewStructure/useGetFullReviewStructureAsync.tsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    variables: {
+      reviewAssignmentId: props.reviewAssignment.id as number,
+      sectionIds,
+      userId: currentUser?.userId as number,
+    },
+    fetchPolicy: 'network-only',
+    skip: !awaitModeResolver,
+  })
+
+  console.log(awaitModeResolver, !!awaitModeResolver)
```
[the remark this answers is not in the document]
Comment (andreievg) [reply src/utils/hooks/useGetFullReviewStructure/useGetFullReviewStructureAsync.tsx]: 9ec663fbc4933104606e11262afcb25c286c7a92

**Comment by andreievg** [comment]
> Just made comments to remove prints so far. I think this one would be better to review after the merge with master is done, but after trying to do it it doesn't seem very easy without having to understand exactly the logic used for this. So I'll leave that until Andrei returns.

> Pending: resolve conflicts with master and do more tests (I've done tests on it before the merge and seems good)

@nmadruga I've updated from master

**Comment by nmadruga** [comment]
Thanks!

Also I think this is linking to the wrong issue (it's pointing to another PR - is that correct?):

> Fixes: #468

Maybe that's part of issue #442? "Part2 - show previous review-response? fix bug with review summary changing when application re-submission is started, looks like draft application responses are mixing things up"

**Comment by nmadruga** [comment]
I tested with master logged as **Reviewer1** and application serial **23456**, because this user can review both sections.
All seemed fine with this now (using `master` back-end`), after LOQ was send and Applicant re-submitted this is how it's looking:
Screen Shot 2021-04-08 at 12 03 45 PM
Screen Shot 2021-04-08 at 12 04 01 PM
Screen Shot 2021-04-08 at 12 04 24 PM

**Review by nmadruga — APPROVED** [review]
LGTM

Please: Just update to the correct issue number on the title and description so we can refer to that later if needed!

**Comment by andreievg** [comment]
> Please: Just update to the correct issue number on the title and description so we can refer to that later if needed! 

done

**Comment by andreievg** [comment]
thanks for review

**Pull request merged**